keps/sig-node/24-apparmor/README.md

```diff
@@ -0,0 +1,790 @@
+
+ * `apparmor.security.beta.kubernetes.io/defaultProfileName`
+ * `apparmor.security.beta.kubernetes.io/allowedProfileNames`
+
+To provide this functionality, users will need to implement equivalent rules in admission control,
+by injecting or validating a `LocalHostProfile` field for Pods or individual containers.
+
+#### Pod Update
+
+The AppArmor fields on a pod are immutable, which also applies to the
+[annotation](https://github.com/kubernetes/kubernetes/blob/b46612a74224b0871a97dae819f5fb3a1763d0b9/pkg/apis/core/validation/validation.go#L177-L182).
+
+When an [Ephemeral Container](20190212-ephemeral-containers.md) is added, it
+will follow the same rules for using or overriding the pod's AppArmor profile.
+Ephemeral container's will never sync with an AppArmor annotation.
+
+#### PodTemplates
+
+PodTemplates (e.g. ReplaceSets, Deployments, StatefulSets, etc.) will be
+ignored. The field/annotation resolution will happen on template instantiation.
+
+To raise awareness of existing controllers using the AppArmor annotations that
+need to be migrated, a warning mechanism will be used to highlight that support
+will be dropped in v1.30.
+
+The mechanisms being considered are audit annotations, annotations on the
+object, events, or a warning as described in [KEP
+#1693](/keps/sig-api-machinery/1693-warnings).
+
+#### Runtime Profiles
+
+The API Server will continue to reject annotations with runtime profiles
+different than `runtime/default`, to maintain the existing behavior.
+
+Violations would lead to the error message:
+
+```
+Invalid value: "runtime/profile-name": must be a valid AppArmor profile
+```
+
+#### Kubelet Backwards compatibility
+
+The changes brought to the Kubelet by this KEP will ensure backwards
+compatibility in a similar way the changes above define it at API Server level.
+Therefore, the AppArmor profiles will be applied following the priority order:
+
+1. Container-specific field.
+2. Container-specific annotation.
+3. Pod-wide field.
+
+In case annotations and fields at either container or pod level exist, the
+kubelet will ignore the annotations and will only apply the profile defined on
+the relevant field.
+
+#### Upgrade / Downgrade
+
+Nodes do not currently support in-place upgrades, so pods will be recreated on
+node upgrade and downgrade. No special handling or consideration is needed to
+support this.
+
+On the API server side, we've already taken version skew in HA clusters into
+account. The same precautions make upgrade & downgrade handling a non-issue.
+
+Since
+[we support](https://kubernetes.io/docs/setup/release/version-skew-policy/) up
+to 2 minor releases of version skew between the master and node, annotations
+must continue to be supported and backfilled for at least 2 versions passed the
+initial implementation. However, we can decide to extend support farther to
+reduce breakage. If this feature is implemented in v1.26, I propose v1.30 as a
+target for removal of the old behavior.
```

Comment:
With seccomp, after the version skew period, we decided to:

1. Remove annotation support in the Kubelet
2. Stop copying fields -> annotatinos

However, we decided to continue validating annotations, and continue copying annotations -> fields indefinitely.

We should follow the same pattern here.